File: sorting functions/quicksort.js

let  arr = [100, 20, 50, 50, 70, 80, 70, 60, 10, 130];
arr = arr.map(x => x*x);
console.log(arr)
let sorted = quicksort(arr,true);
console.log(sorted)



function quicksort(arr, reverse = false){
   arr = partition(arr, 0, arr.length-1, reverse);
    return arr;
}

function partition(arr, low, high, reverse){
    let pivot = arr[high];
    let position = low;

    if (!reverse){
        for (i = low; i < high; i++){
            if(arr[i] < pivot){
                arr = swap(arr, i, position);
                position++;
            }
        }
    }
    else{
        for (i = low; i < high; i++){
            if(arr[i] > pivot){
                arr = swap(arr, i, position);
                position++;
            }
        }
    }

    arr = swap(arr, position, high);

    if(position - 1 > low){
        partition(arr, low, position - 1,reverse);
    }
    if(position < high){
        partition(arr, position + 1, high,reverse);
    }

    return arr;
    
}

function swap(arr, p1, p2){
    let temp = arr[p1]; 
    arr[p1] = arr[p2];
    arr[p2] = temp;
    return arr;
}

File: sorting functions/quicksort.java

import java.util.Arrays;


public class quicksort{
    public static void main(String[] args){
        int[] arr = {80,90,30,50,70,60,100,60,120, 70, 130, 150};

        
        String[] str_arr = Arrays.stream(arr).mapToObj(String::valueOf).toArray(String[]::new);
        System.out.printf(String.join(",", str_arr) +"\n");
        
        qs(arr,true);

        str_arr = Arrays.stream(arr).mapToObj(String::valueOf).toArray(String[]::new);
        System.out.printf(String.join(",", str_arr) +"\n");
  
    }


    public static void qs(int[] arr, boolean reverse){
     part(arr, 0, arr.length-1, true);
    }

    public static void part(int[] arr, int low, int high, boolean reverse){
        int placeholder = low;
        
        int pivot = arr[high];

        for(int i = low; i < high; i++){
            if(reverse){
                if(arr[i] >= pivot){
                    swap(arr,placeholder,i);
                    placeholder++;
                }
            } 
            else{
                if(arr[i] <= pivot){
                    swap(arr,placeholder,i);
                    placeholder++;
                }
            }    
        }
        swap(arr,placeholder, high);

        if (placeholder != low){
            part(arr, 0, placeholder - 1, reverse);
        }
        if (placeholder != high){
            part(arr, placeholder, high, reverse);
        }   
    }

    public static void swap(int[] arr, int p1, int p2){
        int temp = arr[p1];
        arr[p1] = arr[p2];
        arr[p2] = temp;
    }
}
